File: 147.对链表进行插入排序.js
/*
 * @lc app=leetcode.cn id=147 lang=javascript
 *
 * [147] 对链表进行插入排序
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var insertionSortList = function(head) {

    if (!head) {
        return head;
    }

    let dummp = new ListNode();
    dummp.next = head;
    let last = head;
    let cur = head.next;

    while (cur) {
        if (cur.val > last.val) {
            last = last.next;
            cur = cur.next;
        } else {
            let p = dummp;
            while (p.next.val < cur.val) {
                p = p.next;
            }
            last.next = cur.next;
            cur.next = p.next;
            p.next = cur;
            cur = last.next;
        }
    }

    return dummp.next;

};
// @lc code=end

File: 147.对链表进行插入排序.java
/*
 * @lc app=leetcode.cn id=147 lang=java
 *
 * [147] 对链表进行插入排序
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode insertionSortList(ListNode head) {
        if (head == null) {
            return head;
        }

        ListNode dummp = new ListNode();
        dummp.next = head;
        ListNode last = head;
        ListNode cur = head.next;

        while (cur != null) {
            if (cur.val > last.val) {
                last = last.next;
                cur = cur.next;
            } else {
                ListNode pre = dummp;
                while (pre.next.val < cur.val) {
                    pre = pre.next;
                }
                last.next = cur.next;
                cur.next = pre.next;
                pre.next = cur;
                cur = last.next;
            }
        }

        return dummp.next;

    }
}
// @lc code=end

